src/uct/ib/mlx5/ib_mlx5.c

```diff
@@ -700,3 +715,119 @@ void uct_ib_mlx5_destroy_qp(uct_ib_mlx5_md_t *md, uct_ib_mlx5_qp_t *qp)
+        /* have SLs with AR, and AR is YES/TRY */
+        sl = ucs_ffs64_safe(sls_with_ar);
+        ucs_assert(sl != 64);
+
+        sl_ar_support_str = "yes";
+    } else if (((ar_enable == UCS_NO) ||
+                (ar_enable == UCS_TRY)) && (sls_without_ar != 0)) {
+        /* have SLs without AR, and AR is NO/TRY */
+        sl = ucs_ffs64_safe(sls_without_ar);
+        ucs_assert(sl != 64);
+
+        sl_ar_support_str = "no";
+    } else if (ar_enable == UCS_TRY) {
+        sl = ucs_ffs64_safe(sl_allow_mask);
+        ucs_assert(sl != 64);
+
+        sl_ar_support_str = "unknown";
+    } else {
+        sl_ar_support_str = (ar_enable == UCS_YES) ? "with" : "without";
+        goto err;
+    }
+
+    *sl_p = sl;
+    ucs_debug("SL=%u (AR support - %s) was selected on %s,"
+              " SLs with AR support = { %s }, SLs without AR support = { %s }",
+              sl, sl_ar_support_str, dev_port_name,
+              ucs_mask_str(sls_with_ar, &sls_with_ar_str),
+              ucs_mask_str(sls_without_ar, &sls_without_ar_str));
+out_str_buf_clean:
+    ucs_string_buffer_cleanup(&sls_with_ar_str);
+    ucs_string_buffer_cleanup(&sls_without_ar_str);
+    return status;
+
+err:
+    ucs_assert(ar_enable != UCS_TRY);
+    ucs_config_sprintf_ulunits(sl_str, sizeof(sl_str), &ib_config->sl, NULL);
+    ucs_config_sprintf_ternary(ar_enable_str, sizeof(ar_enable_str), &ar_enable,
+                               NULL);
+    ucs_error("AR=%s was requested for SL=%s, but %s %s AR on %s,"
+              " SLs with AR support = { %s }, SLs without AR support = { %s }",
+              ar_enable_str, sl_str,
+              have_sl_mask_cap ? "could not select SL" :
+              "could not detect AR mask for SLs. Please, set SL manually",
+              sl_ar_support_str, dev_port_name,
+              ucs_mask_str(sls_with_ar, &sls_with_ar_str),
+              ucs_mask_str(sls_without_ar, &sls_without_ar_str));
+    status = UCS_ERR_UNSUPPORTED;
+    goto out_str_buf_clean;
+}
+
+ucs_status_t
+uct_ib_mlx5_iface_select_sl(const uct_ib_mlx5_iface_config_t *ib_mlx5_config,
+                            const uct_ib_iface_config_t *ib_config,
+                            const uct_iface_params_t *params,
+                            uct_ib_mlx5_md_t *md, uint8_t *sl_p)
+{
+    uint16_t ooo_sl_mask = 0;
+    ucs_status_t status;
+    uint8_t port_num;
+
+    status = uct_ib_device_find_port(&md->super.dev,
+                                     params->mode.device.dev_name,
+                                     &port_num);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    status = UCS_ERR_UNSUPPORTED;
+#if HAVE_DEVX
+    if (ib_mlx5_config->ar_enable != UCS_NO) {
+        status = uct_ib_mlx5_devx_query_ooo_sl_mask(md, port_num,
+                                                    &ooo_sl_mask);
+        if ((status != UCS_OK) && (status != UCS_ERR_UNSUPPORTED)) {
+            return status;
+        }
+    }
+#endif
```

Comment:
1. can we unite uct_ib_mlx5_select_sl and uct_ib_mlx5_iface_select_sl?
2. why need to check ib_mlx5_config->ar_enable==NO? if user forces no-AR, we still want to know SLs..